# Multi Linear Regression

## $$Y = b_0 + b_1X_1 + b_2X_2 + b_3X_3 + b_4X_4$$

### Predicted Yield = Base + (Coefficient₁ × Urea) + (Coefficient₂ × Pest) + (Coefficient₃ × Rainfall) + (Coefficient₄ × Max_Temp)

In [4]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import r2_score, mean_squared_error, mean_absolute_error
print("libraries imported")



libraries imported


In [7]:
df = pd.read_csv("paddydataset.csv")
print(df.shape)
print(df.columns)

(2789, 45)
Index(['Hectares ', 'Agriblock', 'Variety', 'Soil Types', 'Seedrate(in Kg)',
       'LP_Mainfield(in Tonnes)', 'Nursery', 'Nursery area (Cents)',
       'LP_nurseryarea(in Tonnes)', 'DAP_20days', 'Weed28D_thiobencarb',
       'Urea_40Days', 'Potassh_50Days', 'Micronutrients_70Days',
       'Pest_60Day(in ml)', '30DRain( in mm)', '30DAI(in mm)',
       '30_50DRain( in mm)', '30_50DAI(in mm)', '51_70DRain(in mm)',
       '51_70AI(in mm)', '71_105DRain(in mm)', '71_105DAI(in mm)',
       'Min temp_D1_D30', 'Max temp_D1_D30', 'Min temp_D31_D60',
       'Max temp_D31_D60', 'Min temp_D61_D90', 'Max temp_D61_D90',
       'Min temp_D91_D120', 'Max temp_D91_D120',
       'Inst Wind Speed_D1_D30(in Knots)', 'Inst Wind Speed_D31_D60(in Knots)',
       'Inst Wind Speed_D61_D90(in Knots)',
       'Inst Wind Speed_D91_D120(in Knots)', 'Wind Direction_D1_D30',
       'Wind Direction_D31_D60', 'Wind Direction_D61_D90',
       'Wind Direction_D91_D120', 'Relative Humidity_D1_D30',
       'Re

x1 = 'Urea_40Days', x2 = 'Pest_60Day(in ml)', x3 = '30_50DRain( in mm)', x4 = 'Max temp_D31_D60'
y cap = 'Paddy yield(in Kg)'

In [8]:
## Features selected
features_selected = ["Urea_40Days", "Pest_60Day(in ml)", "30_50DRain( in mm)", "Max temp_D31_D60"]
target = 'Paddy yield(in Kg)'
X = df[features_selected].copy()
y = df[target].copy()
print("features and target variable selected")
print("features selected:", features_selected)
print(X)
print(y)


features and target variable selected
features selected: ['Urea_40Days', 'Pest_60Day(in ml)', '30_50DRain( in mm)', 'Max temp_D31_D60']
      Urea_40Days  Pest_60Day(in ml)  30_50DRain( in mm)  Max temp_D31_D60
0          162.78               3600               187.2                30
1          162.78               3600               187.2                35
2          162.78               3600               185.2                30
3          162.78               3600               185.2                32
4          162.78               3600               185.6                28
...           ...                ...                 ...               ...
2784        27.13                600               185.2                32
2785        27.13                600               185.6                28
2786        27.13                600               185.6                34
2787        27.13                600               187.2                30
2788        27.13                600   

In [9]:
print(X.shape)
print(y.shape)

(2789, 4)
(2789,)


In [10]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)
print("length of x train:", len(X_train))
print("length of x test:", len(X_test))
print("length of y train:", len(y_train))
print("length of y test:", len(y_test))


length of x train: 2231
length of x test: 558
length of y train: 2231
length of y test: 558


In [11]:
model = LinearRegression()
model.fit(X_train, y_train)
print("model trained")
print("y train",y_train)



model trained
y train 599     31780
1378    24724
1717    17649
100     37242
315     30250
        ...  
1638    21968
1095    25244
1130    25228
1294    24908
860     31555
Name: Paddy yield(in Kg), Length: 2231, dtype: int64


### what is model?
A computer program or mathematical algorithm trained on a dataset to identify patterns, make predictions, or generate insights from new, unseen data 

## $$Y = b_0 + b_1X_1 + b_2X_2 + b_3X_3 + b_4X_4$$

In [13]:
Intercept = model.intercept_
Coefficients = model.coef_
print("intercept:", Intercept)
print("coefficients:", Coefficients)

intercept: 1177.274160974317
coefficients: [  0.47869386  10.5866684  -11.60218195  -4.97785833]


## $$Y = b_0 + b_1X_1 + b_2X_2 + b_3X_3 + b_4X_4$$

predicted yield = 1177.27 + 0.47*Urea_40Days + 10.5*Pest_60Day(in ml) -11.60*30_50DRain( in mm) -4.97*Max temp_D31_D60

In [13]:
# STEP 5: Create and Train the Model
# This is where the MAGIC happens! 🎩✨

model = LinearRegression()
model.fit(X_train, y_train)

print("✅ MODEL TRAINED!")
print("\n" + "="*60)
print("🎯 THE EQUATION OUR MODEL LEARNED:")
print("="*60)

# Extract the coefficients (the "factors" we talked about)
intercept = model.intercept_
coefficients = model.coef_

print(f"\n🏠 Base Yield (Intercept): {intercept:,.2f} Kg")
print(f"\n📐 Feature Coefficients (Impact Factors):")
for feature, coef in zip(features_selected, coefficients):
    print(f"  • {feature:30s} → {coef:+.4f}")

print("\n" + "="*60)
print("📝 PRACTICAL EQUATION:")
print("="*60)
print(f"\nPredicted_Yield = {intercept:,.2f}")
for feature, coef in zip(features_selected, coefficients):
    print(f"                 {coef:+.4f} × {feature}")
    
print("\n" + "-"*60)
print("WHAT THIS MEANS:")
print("-"*60)
for feature, coef in zip(features_selected, coefficients):
    unit = "Kg" if "Urea" in feature else ("ml" if "Pest" in feature else ("mm" if "Rain" in feature else "°C"))
    impact = "increases" if coef > 0 else "decreases"
    print(f"  • 1 {unit} more {feature.split('(')[0]:<20} → Yield {impact} by {abs(coef):.2f} Kg")

✅ MODEL TRAINED!

🎯 THE EQUATION OUR MODEL LEARNED:

🏠 Base Yield (Intercept): 1,177.27 Kg

📐 Feature Coefficients (Impact Factors):
  • Urea_40Days                    → +0.4787
  • Pest_60Day(in ml)              → +10.5867
  • 30_50DRain( in mm)             → -11.6022
  • Max temp_D31_D60               → -4.9779

📝 PRACTICAL EQUATION:

Predicted_Yield = 1,177.27
                 +0.4787 × Urea_40Days
                 +10.5867 × Pest_60Day(in ml)
                 -11.6022 × 30_50DRain( in mm)
                 -4.9779 × Max temp_D31_D60

------------------------------------------------------------
WHAT THIS MEANS:
------------------------------------------------------------
  • 1 Kg more Urea_40Days          → Yield increases by 0.48 Kg
  • 1 ml more Pest_60Day           → Yield increases by 10.59 Kg
  • 1 mm more 30_50DRain           → Yield decreases by 11.60 Kg
  • 1 °C more Max temp_D31_D60     → Yield decreases by 4.98 Kg


In [14]:
# STEP 6: Make Predictions (The Proof in Action!)
print("\n" + "="*70)
print("🔮 MAKING PREDICTIONS ON TEST DATA")
print("="*70)

y_pred_train = model.predict(X_train)
y_pred_test = model.predict(X_test)

# Show some real examples
print("\n📊 SAMPLE PREDICTIONS (First 10 test cases):\n")
comparison = pd.DataFrame({
    'Feature 1\n(Urea)': X_test[features_selected[0]].values[:10],
    'Feature 2\n(Pest)': X_test[features_selected[1]].values[:10],
    'Feature 3\n(Rain)': X_test[features_selected[2]].values[:10],
    'Feature 4\n(Temp)': X_test[features_selected[3]].values[:10],
    'Actual Yield\n(Kg)': y_test.values[:10],
    'Predicted Yield\n(Kg)': y_pred_test[:10],
    'Error\n(Kg)': y_test.values[:10] - y_pred_test[:10]
})
print(comparison.to_string())
print(f"\n{'─'*70}")


🔮 MAKING PREDICTIONS ON TEST DATA

📊 SAMPLE PREDICTIONS (First 10 test cases):

   Feature 1\n(Urea)  Feature 2\n(Pest)  Feature 3\n(Rain)  Feature 4\n(Temp)  Actual Yield\n(Kg)  Predicted Yield\n(Kg)  Error\n(Kg)
0             108.52               2400              185.2                 30               25640           24339.166326  1300.833674
1             108.52               2400              185.2                 32               22512           24329.210609 -1817.210609
2             108.52               2400              187.2                 35               25228           24291.072670   936.927330
3             135.65               3000              185.6                 34               30575           30679.602023  -104.602023
4              81.39               1800              187.2                 30               17649           17950.973959  -301.973959
5             162.78               3600              185.2                 32               36690           37059.1

In [22]:
print("model performance evaluation")
r2_score_train = r2_score(y_train, y_pred_train)
r2_score_test = r2_score(y_test, y_pred_test)
print(f"R² Score (Train): {r2_score_train:.4f}")    
print(f"R² Score (Test): {r2_score_test:.4f}")

mse = mean_squared_error(y_test, y_pred_test)
print(f"Mean Squared Error (MSE): {mse:.4f}")
rmse = np.sqrt(mse)
print(f"Root Mean Squared Error (RMSE): {rmse:.4f}")
mse = mean_squared_error(y_train, y_pred_train)
print(f"Mean Squared Error for y_train and y_test: {mse:.4f}")
rmse = np.sqrt(mse)
print(f"Root Mean Squared Error for y_train and y_test: {rmse:.4f}")

model performance evaluation
R² Score (Train): 0.9894
R² Score (Test): 0.9889
Mean Squared Error (MSE): 901737.2521
Root Mean Squared Error (RMSE): 949.5985
Mean Squared Error for y_train and y_test: 903555.3338
Root Mean Squared Error for y_train and y_test: 950.5553


In [16]:
rmse = np.sqrt(mse)
print(f"Root Mean Squared Error (RMSE): {rmse:.4f}")

Root Mean Squared Error (RMSE): 949.5985
